zebra-chain/src/orchard_zsa/issuance.rs

```diff
@@ -0,0 +1,260 @@
+    mut reader: R,
+    rho: &Rho,
+) -> Result<RandomSeed, SerializationError> {
+    Option::from(RandomSeed::from_bytes(reader.read_32_bytes()?, rho))
+        .ok_or_else(|| SerializationError::Parse("Invalid orchard_zsa RandomSeed!"))
+}
+
+impl ZcashSerialize for NoteValue {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        // FIXME: use Amount serializer/deserializer?
+        writer.write_u64::<LittleEndian>(self.inner())?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for NoteValue {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Ok(NoteValue::from_raw(reader.read_u64::<LittleEndian>()?))
+    }
+}
+
+impl ZcashSerialize for Note {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.recipient().zcash_serialize(&mut writer)?;
+        self.value().zcash_serialize(&mut writer)?;
+        self.asset().zcash_serialize(&mut writer)?;
+        self.rho().zcash_serialize(&mut writer)?;
+        self.rseed().zcash_serialize(&mut writer)?;
+
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for Note {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let recipient = (&mut reader).zcash_deserialize_into()?;
+        let value = (&mut reader).zcash_deserialize_into()?;
+        let asset = (&mut reader).zcash_deserialize_into()?;
+        let rho = (&mut reader).zcash_deserialize_into()?;
+        let rseed = zcash_deserialize_random_seed(&mut reader, &rho)?;
+
+        Option::from(Note::from_parts(recipient, value, asset, rho, rseed))
+            .ok_or_else(|| SerializationError::Parse("Invalid orchard_zsa Note components!"))
+    }
+}
+
+impl TrustedPreallocate for Note {
+    fn max_allocation() -> u64 {
+        // FIXME: is this a correct calculation way?
+        // The longest Vec<Note> we receive from an honest peer must fit inside a valid block.
+        // Since encoding the length of the vec takes at least one byte, we use MAX_BLOCK_BYTES - 1
+        (MAX_BLOCK_BYTES - 1) / NOTE_SIZE
+    }
+}
+
+impl ZcashSerialize for IssueAction {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u8(self.is_finalized().as_u8())?;
+        self.notes().zcash_serialize(&mut writer)?;
+        self.asset_desc().zcash_serialize(&mut writer)?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for IssueAction {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let finalize = reader.read_u8()?.as_bool();
+        let notes = (&mut reader).zcash_deserialize_into()?;
+        let asset_descr = (&mut reader).zcash_deserialize_into()?;
+        Ok(IssueAction::from_parts(asset_descr, notes, finalize))
+    }
+}
+
+impl TrustedPreallocate for IssueAction {
+    fn max_allocation() -> u64 {
+        // FIXME: impl correct calculation
+        10
```

Comment:
This looks good, the function that calls `max_allocation()` ([`zcash_deserialize_external_count()`](https://github.com/QED-it/zebra/blob/switch-to-zsa-crates-nu6-txv6-gen/zebra-chain/src/serialization/zcash_deserialize.rs#L79-L100)) will deserialize and pre-allocate sequentially based on the actual counts, so until it goes through each item and reads their `CompactSize`, it'll just be pre-allocating memory for the pointers, if one of them is shorter than expected it'll return an error, and their combined real size can't be longer than the `Codec`'s `builder.max_len` field, which is `MAX_PROTOCOL_MESSAGE_LEN`.

```suggestion
        (MAX_BLOCK_BYTES - 1) / NOTE_SIZE
```